sql/src/main/java/io/crate/planner/consumer/SemiJoins.java

```diff
@@ -0,0 +1,259 @@
+
+        removeRewriteCandidatesFromWhere(rel, rewriteCandidate);
+        QuerySpec newQS = rel.querySpec().copyAndReplace(refsToFields);
+
+        // Avoid name clashes if the subquery is on the same relation; e.g.: select * from t1 where x in (select * from t1)
+        QualifiedName subQueryName = selectSymbol.relation().getQualifiedName().withPrefix("S");
+
+        // Using MSS instead of TwoTableJoin so that the "fetch-pushdown" logic in the Planner is also applied
+        HashMap<QualifiedName, AnalyzedRelation> sources = new LinkedHashMap<>(2);
+        sources.put(rel.getQualifiedName(), sourceRel);
+        sources.put(subQueryName, selectSymbol.relation());
+
+        // normalize is done to rewrite  SELECT * from t1, t2 to SELECT * from (select ... t1) t1, (select ... t2) t2
+        // because planner logic expects QueriedRelation in the sources
+        MultiSourceSelect mss = new MultiSourceSelect(
+            sources,
+            rel.fields(),
+            newQS,
+            new ArrayList<>(Collections.singletonList(
+                JoinPair.of(
+                    rel.getQualifiedName(),
+                    subQueryName,
+                    JoinType.SEMI,
+                    joinCondition
+                )))
+        );
+        return (QueriedRelation) relationNormalizer.normalize(mss, transactionCtx);
+    }
+
+    @Nullable
+    private static AnalyzedRelation getSource(QueriedRelation rel) {
+        if (rel instanceof QueriedTableRelation) {
+            return ((QueriedTableRelation) rel).tableRelation();
+        }
+        // TODO: support other cases as well
+        return null;
+    }
+
+    private static void removeRewriteCandidatesFromWhere(QueriedRelation rel, Function rewriteCandidate) {
+        rel.querySpec().where().replace(FuncSymbols.mapNodes(f -> {
+            if (f == rewriteCandidate) {
+                return Literal.BOOLEAN_TRUE;
+            }
+            return f;
+        }));
+    }
+
+    static List<Function> gatherRewriteCandidates(Symbol query) {
+        ArrayList<Function> candidates = new ArrayList<>();
+        RewriteCandidateGatherer.INSTANCE.process(query, candidates);
+        return candidates;
+    }
+
+    @Nullable
+    static SelectSymbol getSubqueryOrNull(Symbol symbol) {
+        symbol = unwrapCast(symbol);
+        if (symbol instanceof SelectSymbol) {
+            return ((SelectSymbol) symbol);
+        }
+        return null;
+    }
+
+    private static Symbol unwrapCast(Symbol symbol) {
+        while (isCastFunction(symbol)) {
+            symbol = ((Function) symbol).arguments().get(0);
+        }
+        return symbol;
+    }
+
+    /**
+     * t1.x IN (select y from t2)  --> SEMI JOIN t1 on t1.x = t2.y
+     */
+    static Symbol makeJoinCondition(Function rewriteCandidate, AnalyzedRelation sourceRel, QueriedRelation subRel) {
+        assert getSubqueryOrNull(rewriteCandidate.arguments().get(1)).relation() == subRel
```

Comment:
what about null?

Reply:
Yeah it's a bit nasty to not check for null, but I wanted to avoid introducing a local variable just for the assertion.

I was also thinking about returning a custom data structore from `gatherRewriteCandidates` to avoid having to re-resolve the subQuery/relation. That would avoid the assertions here entirely.

I'll push another commit with that to see if it's cleaner